File: app/static/query-planner.js
let plannerState = null;
let plannerRow = 0;
const plannerOperators = {
  "=": "equals",
  "<>": "not equal",
  begins_with: "starts with",
  between: "between",
  "<": "less than",
  "<=": "at most",
  ">": "greater than",
  ">=": "at least",
  exists: "exists",
  not_exists: "does not exist",
};
function renderQueryPlanner() {
  const table = state.table;
  if (plannerState?.table !== table)
    plannerState = { table, result: null, input: null };
  const fields = state.detail.AttributeDefinitions.map(
    (a) => `<option value="${esc(a.AttributeName)}"></option>`,
  ).join("");
  const last = state.lastRead?.table === table ? state.lastRead : null;
  const discard = last?.scanned
    ? Math.round(100 * (1 - last.count / last.scanned))
    : null;
  document.getElementById("detail-content").innerHTML =
    `<section class="panel"><div class="panel-heading"><div><h2>Access-pattern query builder</h2><p>Describe what you need. Compare the table and every index before reading data.</p></div>${icon("search")}</div><form id="planner-form" class="panel-body"><div class="planner-intro"><label class="field" for="planner-template">Start from a pattern<select id="planner-template"><option value="custom">Custom lookup</option><option value="collection">Single-table entity collection</option><option value="sparse">Sparse tenant work queue</option><option value="inverted">Inverted relationship lookup</option></select></label><button class="button" type="button" id="planner-current">Use current Items query</button></div><p class="info-note" id="planner-pattern-help">Conditions are combined with AND. Attribute names are literal, including dots. No reads run when you compare plans.</p><datalist id="planner-attributes">${fields}</datalist><div id="planner-conditions"></div><button class="button small" type="button" id="planner-add">${icon("plus")}Add condition</button><div class="field-pair mt"><div class="field"><label for="planner-projection">Returned attributes · one per line</label><textarea id="planner-projection" rows="3" placeholder="Blank requires all table attributes"></textarea></div><div><div class="field"><label for="planner-order">Required result order (optional attribute)</label><input id="planner-order" list="planner-attributes" placeholder="e.g. createdAt"></div><label class="planner-check"><input type="checkbox" id="planner-descending">Descending order</label></div></div><label class="planner-check"><input type="checkbox" id="planner-consistent">Require strongly consistent reads</label><label class="planner-check"><input type="checkbox" id="planner-sparse">Allow indexed-items-only results when an index can omit matching items</label><details class="type-guide"><summary>Propose an index design</summary><p>Choose attributes from your conditions. The partition attribute must use equality. A proposal is an editable design, not a schema change.</p><div class="field-pair"><div class="field"><label for="planner-draft-pk">Proposed partition attribute</label><input id="planner-draft-pk" list="planner-attributes" placeholder="e.g. openCustomerKey"></div><div class="field"><label for="planner-draft-sk">Proposed sort attribute (optional)</label><input id="planner-draft-sk" list="planner-attributes" placeholder="e.g. createdAt"></div></div></details><p id="planner-error" class="form-error" role="alert"></p><button class="button primary" type="submit">${icon("search")}Compare access paths</button></form></section>${last ? `<section class="panel mt"><div class="panel-heading"><div><h2>Last measured read</h2><p>${esc(last.index || "Base table")} · ${esc(last.mode)} · one response page</p></div></div><div class="panel-body"><div class="model-summary"><strong>${fmt(last.scanned)} evaluated</strong><span>${fmt(last.count)} returned</span><span>${last.capacity} RCU</span><span>${discard === null ? "No evaluated items" : discard + "% discarded by server filters"}</span></div><p class="info-note">${discard > 50 ? "Many evaluated items were discarded. Look for a plan that moves a frequent filter into the key condition. " : ""}This measures the last read only. Page filters in the UI are separate. It does not predict savings or identify hot partitions.</p></div></section>` : ""}<div id="planner-results" aria-live="polite"></div>`;
  const form = document.getElementById("planner-form");
  form.onsubmit = compareQueryPlans;
  document.getElementById("planner-add").onclick = () => {
    addPlannerCondition();
    form.dispatchEvent(new Event("input"));
  };
  document.getElementById("planner-current").onclick = () => {
    try {
      fillPlannerFromItems();
    } catch (error) {
      document.getElementById("planner-error").textContent = error.message;
    }
  };
  document.getElementById("planner-template").onchange = applyPlannerTemplate;
  form.addEventListener("input", () => {
    plannerState.input = readPlanner();
    plannerState.result = null;
    document.getElementById("planner-results").innerHTML =
      '<p class="info-note">Pattern changed. Compare again to refresh the plans.</p>';
  });
  if (plannerState.input) fillPlanner(plannerState.input);
  else
    addPlannerCondition({
      attribute: state.detail.KeySchema.find((k) => k.KeyType === "HASH")
        .AttributeName,
    });
}
function addPlannerCondition(rule = {}) {
  const host = document.getElementById("planner-conditions");
  if (host.children.length >= 20) return;
  const id = ++plannerRow;
  const row = document.createElement("div");
  row.className = "planner-condition";
  row.innerHTML = `<div class="field"><label for="plan-attr-${id}">Attribute</label><input id="plan-attr-${id}" class="plan-attribute" list="planner-attributes" required maxlength="255" value="${esc(rule.attribute || "")}"></div><div class="field"><label for="plan-op-${id}">Condition</label><select id="plan-op-${id}" class="plan-operator">${Object.entries(
    plannerOperators,
  )
    .map(([v, l]) => `<option value="${esc(v)}">${esc(l)}</option>`)
    .join(
      "",
    )}</select></div><div class="field plan-value-field"><label for="plan-type-${id}">Type</label><select id="plan-type-${id}" class="plan-type"><option value="S">String</option><option value="N">Number</option><option value="B">Binary (base64)</option></select></div><div class="field plan-value-field"><label for="plan-value-${id}">Value</label><input id="plan-value-${id}" class="plan-value" maxlength="4096" value="${esc(rule.value ? Object.values(rule.value)[0] : "")}"></div><div class="field plan-end-field"><label for="plan-end-${id}">Range end</label><input id="plan-end-${id}" class="plan-end" maxlength="4096" value="${esc(rule.end ? Object.values(rule.end)[0] : "")}"></div><button type="button" class="icon-button" aria-label="Remove condition ${id}">${icon("close")}</button>`;
  row.querySelector(".plan-operator").value = rule.operator || "=";
  row.querySelector(".plan-type").value = rule.value
    ? Object.keys(rule.value)[0]
    : state.detail.AttributeDefinitions.find(
        (a) => a.AttributeName === rule.attribute,
      )?.AttributeType || "S";
  const update = () => {
    const op = row.querySelector(".plan-operator").value;
    row
      .querySelectorAll(".plan-value-field")
      .forEach((el) => (el.hidden = ["exists", "not_exists"].includes(op)));
    row.querySelector(".plan-end-field").hidden = op !== "between";
  };
  row.querySelector(".plan-operator").onchange = update;
  row.querySelector("button").onclick = () => {
    row.remove();
    document.getElementById("planner-form").dispatchEvent(new Event("input"));
  };
  update();
  host.append(row);
  if (!rule.attribute) row.querySelector("input").focus();
}
function readPlanner() {
  return {
    conditions: [...document.querySelectorAll(".planner-condition")].map(
      (row) => {
        const rule = {
          attribute: row.querySelector(".plan-attribute").value,
          operator: row.querySelector(".plan-operator").value,
        };
        const kind = row.querySelector(".plan-type").value;
        if (!["exists", "not_exists"].includes(rule.operator))
          rule.value = { [kind]: row.querySelector(".plan-value").value };
        if (rule.operator === "between")
          rule.end = { [kind]: row.querySelector(".plan-end").value };
        return rule;
      },
    ),
    projection: document
      .getElementById("planner-projection")
      .value.split("\n")
      .map((v) => v.trim())
      .filter(Boolean),
    orderBy: document.getElementById("planner-order").value,
    ascending: !document.getElementById("planner-descending").checked,
    consistent: document.getElementById("planner-consistent").checked,
    allowSparse: document.getElementById("planner-sparse").checked,
    draftPartition: document.getElementById("planner-draft-pk").value,
    draftSort: document.getElementById("planner-draft-sk").value,
  };
}
function fillPlanner(input) {
  document.getElementById("planner-conditions").innerHTML = "";
  input.conditions.forEach(addPlannerCondition);
  document.getElementById("planner-projection").value = (
    input.projection || []
  ).join("\n");
  document.getElementById("planner-order").value = input.orderBy || "";
  document.getElementById("planner-consistent").checked =
    input.consistent || false;
  document.getElementById("planner-sparse").checked =
    input.allowSparse || false;
  document.getElementById("planner-descending").checked =
    input.ascending === false;
  document.getElementById("planner-draft-pk").value =
    input.draftPartition || "";
  document.getElementById("planner-draft-sk").value = input.draftSort || "";
  plannerState.input = input;
}
function fillPlannerFromItems() {
  const q = state.search;
  if (q.filterExpression || (q.filterJoin === "OR" && q.filters?.length > 1))
    throw new Error(
      "Advanced expressions and OR conditions need manual review. Enter an equivalent AND access pattern here.",
    );
  const conditions = (q.filters || []).map((rule) => ({ ...rule }));
  if (
    conditions.some(
      (rule) =>
        !plannerOperators[rule.operator] ||
        (rule.value && !["S", "N", "B"].includes(Object.keys(rule.value)[0])),
    )
  )
    throw new Error(
      "This planner supports scalar string, number and binary conditions. Enter a compatible access pattern.",
    );
  if (q.mode === "query") {
    const keys = keySchema(state.detail, q.index);
    if (
      keys.filter((k) => k.KeyType === "HASH").length !== 1 ||
      keys.filter((k) => k.KeyType === "RANGE").length > 1
    )
      throw new Error("Multi-attribute native keys need the advanced AWS API.");
    if (q.partition)
      conditions.unshift({
        attribute: keys.find((k) => k.KeyType === "HASH").AttributeName,
        operator: "=",
        value: q.partition,
      });
    if (q.sort)
      conditions.push({
        attribute: keys.find((k) => k.KeyType === "RANGE").AttributeName,
        operator: q.operator,
        value: q.sort,
        ...(q.sortEnd ? { end: q.sortEnd } : {}),
      });
  }
  if (!conditions.length)
    throw new Error(
      "The current scan has no key or filter conditions. Add the attributes you know for your lookup.",
    );
  fillPlanner({ ...q, conditions });
  document.getElementById("planner-form").dispatchEvent(new Event("input"));
}
function applyPlannerTemplate() {
  const value = document.getElementById("planner-template").value;
  const pk = state.detail.KeySchema.find(
    (k) => k.KeyType === "HASH",
  ).AttributeName;
  const sk =
    state.detail.KeySchema.find((k) => k.KeyType === "RANGE")?.AttributeName ||
    "sk";
  const templates = {
    custom: {
      conditions: [{ attribute: pk, operator: "=", value: { S: "" } }],
    },
    collection: {
      conditions: [
        { attribute: pk, operator: "=", value: { S: "CUSTOMER#123" } },
        {
          attribute: sk,
          operator: "begins_with",
          value: { S: "ORDER#2026-09" },
        },
      ],
    },
    sparse: {
      conditions: [
        {
          attribute: "openCustomerKey",
          operator: "=",
          value: { S: "CUSTOMER#123" },
        },
        {
          attribute: "createdAt",
          operator: "begins_with",
          value: { S: "2026-09" },
        },
      ],
      draftPartition: "openCustomerKey",
      draftSort: "createdAt",
    },
    inverted: {
      conditions: [
        { attribute: sk, operator: "=", value: { S: "PRODUCT#456" } },
        { attribute: pk, operator: "begins_with", value: { S: "CUSTOMER#" } },
      ],
      draftPartition: sk,
      draftSort: pk,
    },
  };
  fillPlanner(templates[value]);
  document.getElementById("planner-pattern-help").textContent = {
    custom:
      "Enter the literal attributes and values for your lookup. All conditions use AND.",
    collection:
      "Example string keys: a customer partition with an ORDER# prefix. Prefixes match from the left; adjust the names and values to your schema.",
    sparse:
      "Example: only open orders carry openCustomerKey and createdAt. Remove openCustomerKey when closed. Scope by customer to avoid one global OPEN partition; measure skew for large customers.",
    inverted:
      "Example: query relationships in reverse by exchanging key roles in a GSI. This requires those attributes and matching types on your relationship items.",
  }[value];
  document.getElementById("planner-form").dispatchEvent(new Event("input"));
}
async function compareQueryPlans(event) {
  event.preventDefault();
  const generation = state.generation,
    current = plannerState;
  const button = event.target.querySelector('[type="submit"]');
  const input = readPlanner();
  current.input = input;
  current.result = null;
  document.getElementById("planner-results").innerHTML =
    '<p class="info-note">Comparing table and index metadata…</p>';
  document.getElementById("planner-error").textContent = "";
  button.disabled = true;
  try {
    const result = await send(
      tablePath(current.table) + "/query-plan",
      "POST",
      input,
    );
    if (
      state.generation !== generation ||
      state.detailTab !== "planner" ||
      plannerState !== current ||
      JSON.stringify(readPlanner()) !== JSON.stringify(input)
    )
      return;
    current.result = result;
    const list = (values) =>
      values.length
        ? `<ul class="planner-notes">${values.map((v) => `<li>${esc(v)}</li>`).join("")}</ul>`
        : "";
    document.getElementById("planner-results").innerHTML =
      `<section class="panel mt"><div class="panel-heading"><div><h2>${result.suggested ? "Compare query plans" : "No supported query covers this pattern"}</h2><p>${esc(result.basis)}</p></div><button class="button small" id="planner-export">${icon("download")}Export plan</button></div><div class="panel-body planner-candidates">${result.candidates.map((c, i) => `<article class="planner-candidate ${c.id === result.suggested ? "planner-suggested" : ""}"><div class="planner-candidate-title"><h3>${esc(c.name)}</h3><span class="tag ${c.blockers.length ? "red" : "green"}">${c.blockers.length ? "Needs changes" : c.id === result.suggested ? "Fewest filters" : "Query available"}</span><span class="tag">${esc(c.kind)}</span></div><p class="info-note">${c.keys.map((k) => `${esc(k.KeyType === "HASH" ? "Partition" : "Sort")}: <code>${esc(k.AttributeName)}</code>`).join(" · ")} · Projection ${esc(c.projection || "—")}</p><p><strong>Key conditions:</strong> ${esc(c.keyConditions.join(", ") || "None")}<br><strong>Post-read filters:</strong> ${esc(c.filterAttributes.join(", ") || "None")}</p>${list(c.blockers)}${list(c.warnings)}${c.request ? `<button class="button primary small" data-plan-use="${i}">Use in Items · review first</button><details class="type-guide"><summary>AWS Query request</summary><pre class="code-block" tabindex="0">${highlightJSON(JSON.stringify(c.awsRequest, null, 2))}</pre></details>` : ""}</article>`).join("")}${result.proposal ? `<article class="planner-candidate"><h3>Proposed GSI design</h3>${list(result.proposal.blockers)}${result.proposal.definition ? `${result.proposal.reuse.length ? `<p class="info-note">Existing indexes already use these keys: ${esc(result.proposal.reuse.join(", "))}. Review reusing them before adding another.</p>` : ""}<pre class="code-block" tabindex="0">${highlightJSON(JSON.stringify(result.proposal.definition, null, 2))}</pre>${list(result.proposal.notes)}` : ""}</article>` : ""}${list(result.notes)}</div></section>`;
    document
      .querySelectorAll("[data-plan-use]")
      .forEach(
        (el) => (el.onclick = () => useQueryPlan(Number(el.dataset.planUse))),
      );
    document.getElementById("planner-export").onclick = () =>
      saveDownload(
        JSON.stringify(
          {
            format: "dynamodb-tools-query-plan",
            version: 1,
            table: current.table,
            connection: workspaceNavigation.scope(state.overview.connection),
            pattern: input,
            ...result,
          },
          null,
          2,
        ),
        current.table + "-query-plan.json",
      ).catch((error) => toast(error.message, true));
  } catch (error) {
    if (state.generation === generation && state.detailTab === "planner") {
      document.getElementById("planner-error").textContent = error.message;
      document.getElementById("planner-results").innerHTML = "";
    }
  } finally {
    button.disabled = false;
  }
}
function useQueryPlan(index) {
  const plan = plannerState?.result?.candidates[index];
  if (!plan?.request || plannerState.table !== state.table) return;
  state.generation++;
  itemRequest++;
  state.busy = false;
  history.replaceState(null, "", tableHash(state.table));
  state.search = structuredClone(plan.request);
  state.savedQueryId = "";
  state.items = [];
  state.page = 1;
  state.cursors = [null];
  state.nextCursor = null;
  state.detailTab = "items";
  document.querySelectorAll(".tab").forEach((el) => {
    const active = el.dataset.tab === "items";
    el.classList.toggle("active", active);
    el.setAttribute("aria-selected", String(active));
    el.tabIndex = active ? 0 : -1;
  });
  document
    .getElementById("detail-content")
    .setAttribute("aria-labelledby", "tab-items");
  renderExplorer();
  document.getElementById("page-label").textContent = "Not run yet";
  document.getElementById("items-container").innerHTML = empty(
    "Query plan ready",
    "Review the conditions, then run or save the query.",
    "",
    true,
  );
  renderWorkspaceNavigation();
  document.getElementById("query-mode").focus();
}

File: app/query_planner.py
"""Metadata-only access-pattern planning. Never executes reads or schema changes."""
import json
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .data_codec import attribute_from_wire


class Condition(BaseModel):
    model_config = ConfigDict(extra="forbid")
    attribute: str = Field(min_length=1, max_length=255)
    operator: Literal[
        "=",
        "<>",
        "<",
        "<=",
        ">",
        ">=",
        "begins_with",
        "between",
        "exists",
        "not_exists",
    ] = "="
    value: dict | None = None
    end: dict | None = None

    @model_validator(mode="after")
    def validate_value(self):
        if self.operator in {"exists", "not_exists"}:
            if self.value is not None or self.end is not None:
                raise ValueError("Existence conditions do not take values")
            return self
        if (
            not self.value
            or len(self.value) != 1
            or next(iter(self.value)) not in {"S", "N", "B"}
        ):
            raise ValueError(
                "Planner values must be a DynamoDB string, number or binary scalar"
            )
        kind = next(iter(self.value))
        attribute_from_wire(self.value)
        if len(json.dumps(self.value).encode()) > 8192:
            raise ValueError("Planner values must be at most 8 KiB")
        if self.operator == "begins_with" and kind == "N":
            raise ValueError("begins_with requires a string or binary value")
        if self.operator == "between":
            if not self.end or set(self.end) != {kind}:
                raise ValueError("A range end with the same type is required")
            lower = attribute_from_wire(self.value)[kind]
            upper = attribute_from_wire(self.end)[kind]
            if len(json.dumps(self.end).encode()) > 8192:
                raise ValueError("Planner values must be at most 8 KiB")
            if kind == "N":
                lower, upper = Decimal(lower), Decimal(upper)
            if lower > upper:
                raise ValueError("Range start must not exceed range end")
        elif self.end is not None:
            raise ValueError("Only BETWEEN takes a range end")
        return self


class PlanRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    conditions: list[Condition] = Field(min_length=1, max_length=20)
    projection: list[str] = Field(default_factory=list, max_length=100)
    consistent: bool = False
    allowSparse: bool = False
    ascending: bool = True
    orderBy: str = Field(default="", max_length=255)
    draftPartition: str = Field(default="", max_length=255)
    draftSort: str = Field(default="", max_length=255)

    @model_validator(mode="after")
    def unique_attributes(self):
        names = [rule.attribute for rule in self.conditions]
        if len(names) != len(set(names)):
            raise ValueError("Use one condition per attribute; use BETWEEN for a range")
        if any(not name or len(name) > 255 for name in self.projection):
            raise ValueError(
                "Returned attributes must be nonempty names up to 255 characters"
            )
        if self.draftPartition and self.draftPartition == self.draftSort:
            raise ValueError("Proposed partition and sort attributes must differ")
        return self


def key_error(rule, types, role):
    if not rule.value:
        return "A key condition needs a value"
    kind, value = next(iter(rule.value.items()))
    if rule.attribute in types and types[rule.attribute] != kind:
        return f"{rule.attribute} requires type {types[rule.attribute]}, not {kind}"
    raw = attribute_from_wire(rule.value)[kind]
    size = len(raw if kind == "B" else value.encode())
    if not size or size > (2048 if role == "HASH" else 1024):
        return f"{rule.attribute} has an empty or oversized key value"
    if rule.end:
        end_rule = Condition(attribute=rule.attribute, value=rule.end)
        return key_error(end_rule, types, role)
    return None


def aws_query(table, request, partition, sort, residual, index):
    names, values = {}, {}

    def expression(rule, prefix):
        alias, token = "#" + prefix, ":" + prefix
        names[alias] = rule.attribute
        if rule.operator in {"exists", "not_exists"}:
            return f"attribute_{rule.operator}({alias})"
        values[token] = rule.value
        if rule.operator == "begins_with":
            return f"begins_with({alias}, {token})"
        if rule.operator == "between":
            values[token + "end"] = rule.end
            return f"{alias} BETWEEN {token} AND {token}end"
        return f"{alias} {rule.operator} {token}"

    key_expression = expression(partition, "pk")
    if sort:
        key_expression += " AND " + expression(sort, "sk")
    result = {
        "TableName": table["TableName"],
        "KeyConditionExpression": key_expression,
        "ScanIndexForward": request.ascending,
        "ConsistentRead": request.consistent,
        "Limit": 25,
        "ReturnConsumedCapacity": "TOTAL",
    }
    if index:
        result["IndexName"] = index
    if residual:
        result["FilterExpression"] = " AND ".join(
            f"({expression(rule, f'f{i}')})" for i, rule in enumerate(residual)
        )
    if request.projection:
        projected = list(
            dict.fromkeys(
                request.projection
                + [key["AttributeName"] for key in table["KeySchema"]]
            )
        )
        for i, name in enumerate(projected):
            names[f"#p{i}"] = name
        result["ProjectionExpression"] = ", ".join(
            f"#p{i}" for i in range(len(projected))
        )
    result["ExpressionAttributeNames"] = names
    result["ExpressionAttributeValues"] = values
    return result


def candidate(table, definition, kind, request):
    rules = {rule.attribute: rule for rule in request.conditions}
    schema = definition["KeySchema"]
    hashes = [k["AttributeName"] for k in schema if k["KeyType"] == "HASH"]
    ranges = [k["AttributeName"] for k in schema if k["KeyType"] == "RANGE"]
    name = definition.get("IndexName")
    result = {
        "id": "index/" + name if name else "__table__",
        "name": name or "Base table",
        "kind": kind,
        "keys": schema,
        "blockers": [],
        "warnings": [],
        "keyConditions": [],
        "filterAttributes": [],
        "request": None,
        "awsRequest": None,
    }
    blockers, warnings = result["blockers"], result["warnings"]
    if len(hashes) != 1 or len(ranges) > 1:
        blockers.append(
            "Multi-attribute keys need the advanced AWS query API; this planner supports one partition and "
            "one sort attribute, including composite strings."
        )
        return result
    pk, sk = hashes[0], ranges[0] if ranges else None
    partition, sort = rules.get(pk), rules.get(sk)
    types = {
        a["AttributeName"]: a["AttributeType"] for a in table["AttributeDefinitions"]
    }
    if not partition or partition.operator != "=":
        blockers.append(
            f"Supply an equality condition on {pk} to use Query. Otherwise this access path needs a scan."
        )
    else:
        error = key_error(partition, types, "HASH")
        if error:
            blockers.append(error)
        result["keyConditions"].append(pk)
    if sort and sort.operator == "exists":
        sort = None  # Every entry already has the index sort key.
    elif sort and sort.operator in {"<>", "not_exists"}:
        blockers.append(
            f"{sk}: this condition cannot be a sort-key condition or a key filter in Query."
        )
    elif sort:
        error = key_error(sort, types, "RANGE")
        if error:
            blockers.append(error)
        result["keyConditions"].append(sk)
    residual = [rule for rule in request.conditions if rule.attribute not in {pk, sk}]
    result["filterAttributes"] = [rule.attribute for rule in residual]
    if residual:
        warnings.append(
            "Remaining conditions are post-read filters. They do not reduce the items evaluated or the read "
            "capacity consumed; pages can be empty and still have a continuation."
        )
    if request.orderBy and request.orderBy != sk:
        blockers.append(
            f"This access path orders by {sk or 'no sort key'}, not {request.orderBy}. "
            "DynamoDB cannot sort by an arbitrary attribute."
        )
    if kind == "GSI" and request.consistent:
        blockers.append(
            "GSIs support eventual consistency only. Use the base table or an LSI for a strongly consistent query."
        )
    if name and definition.get("IndexStatus", "ACTIVE") != "ACTIVE":
        blockers.append("This index is not ACTIVE; wait for it to become available.")
    if kind != "TABLE":
        base_keys = {k["AttributeName"] for k in table["KeySchema"]}
        index_keys = {k["AttributeName"] for k in schema}
        implied = {
            r.attribute for r in request.conditions if r.operator != "not_exists"
        }
        missing_presence = index_keys - base_keys - implied
        warnings.append(
            "Index membership requires every index key attribute to exist. Inspect the Data model sample to "
            "check observed eligibility; samples do not prove complete coverage."
        )
        if missing_presence and not request.allowSparse:
            blockers.append(
                "This index can omit matching items missing "
                + ", ".join(sorted(missing_presence))
                + ". Confirm indexed-items-only results to use this sparse access path."
            )
        projection = definition.get("Projection", {"ProjectionType": "ALL"})
        result["projection"] = projection["ProjectionType"]
        if projection["ProjectionType"] != "ALL":
            available = (
                base_keys | index_keys | set(projection.get("NonKeyAttributes", []))
            )
            needed = set(request.projection) | set(result["filterAttributes"])
            absent = sorted(needed - available)
            if not request.projection:
                blockers.append(
                    "Specify returned attributes: this index does not project every table attribute."
                )
            if absent and kind == "GSI":
                blockers.append(
                    "Not projected into this GSI: "
                    + ", ".join(absent)
                    + ". A GSI cannot fetch these from the base table."
                )
            elif absent:
                warnings.append(
                    "The LSI must fetch unprojected attributes from the base table, adding read cost and latency: "
                    + ", ".join(absent)
                )
        if kind == "GSI":
            warnings.append(
                "A GSI adds storage and write work, may lag the table, and can throttle table writes if its "
                "capacity is insufficient."
            )
    else:
        result["projection"] = "ALL"
    if not blockers:
        result["request"] = {
            "mode": "query",
            "index": name,
            "partition": partition.value,
            "sort": sort.value if sort else None,
            "sortEnd": sort.end if sort else None,
            "operator": sort.operator if sort else "=",
            "ascending": request.ascending,
            "limit": 25,
            "consistent": request.consistent,
            "projection": request.projection,
            "filters": [r.model_dump(exclude_none=True) for r in residual],
            "filterJoin": "AND",
        }
        result["awsRequest"] = aws_query(
            table, request, partition, sort, residual, name
        )
    return result


def proposed_index(table, request):
    if not request.draftPartition:
        return None
    rules = {r.attribute: r for r in request.conditions}
    pk, sk = rules.get(request.draftPartition), rules.get(request.draftSort)
    if not pk or pk.operator != "=":
        return {
            "blockers": ["Choose an equality condition for the proposed partition key."]
        }
    if request.draftSort and (not sk or not sk.value):
        return {"blockers": ["Add a typed condition for the proposed sort attribute."]}
    types = {
        a["AttributeName"]: a["AttributeType"] for a in table["AttributeDefinitions"]
    }
    errors = [
        error
        for rule, role in [(pk, "HASH"), (sk, "RANGE")]
        if rule
        for error in [key_error(rule, types, role)]
        if error
    ]
    if errors:
        return {"blockers": errors}
    schema = [{"AttributeName": pk.attribute, "KeyType": "HASH"}]
    if sk:
        schema.append({"AttributeName": sk.attribute, "KeyType": "RANGE"})
    included = sorted(
        (set(request.projection) | set(rules))
        - {k["AttributeName"] for k in table["KeySchema"] + schema}
    )
    projection = (
        {"ProjectionType": "ALL"}
        if not request.projection
        else {"ProjectionType": "INCLUDE", "NonKeyAttributes": included}
        if included
        else {"ProjectionType": "KEYS_ONLY"}
    )
    if projection["ProjectionType"] == "INCLUDE" and len(included) > 20:
        return {
            "blockers": [
                "More than 20 non-key attributes would need projection. Narrow the returned fields or explicitly "
                "choose all attributes."
            ]
        }
    return {
        "blockers": [],
        "definition": {
            "KeySchema": schema,
            "Projection": projection,
            "AttributeDefinitions": [
                {"AttributeName": r.attribute, "AttributeType": next(iter(r.value))}
                for r in [pk, sk]
                if r
            ],
        },
        "reuse": [
            i["IndexName"]
            for i in table.get("GlobalSecondaryIndexes", [])
            if i["KeySchema"] == schema
        ],
        "notes": [
            "Design proposal only. Choose an index name, capacity settings, quotas and a rollout/backfill "
            "plan before any schema change.",
            "Use high-cardinality partition values; a constant such as OPEN or one popular tenant can "
            "concentrate traffic. Sharding requires querying and merging multiple partitions.",
            "For a sparse index, write dedicated key attributes only on eligible items and remove them when "
            "eligibility ends. A value such as OPEN alone does not make an index sparse.",
            "Application writes must maintain synthetic keys. Existing items need a reviewed migration if "
            "those attributes are absent.",
            "Reuse existing indexes with suitable keys where possible; evaluate all access patterns and "
            "write volume before adding another index.",
            "Projection selection includes filter attributes. Check the aggregate INCLUDE-attribute quota "
            "across indexes, and compare storage/write overhead with measured reads.",
        ],
    }


def plan_access(table, request):
    candidates = [candidate(table, table, "TABLE", request)]
    for kind, field in [
        ("GSI", "GlobalSecondaryIndexes"),
        ("LSI", "LocalSecondaryIndexes"),
    ]:
        candidates.extend(
            candidate(table, index, kind, request) for index in table.get(field, [])
        )
    viable = sorted(
        (c for c in candidates if c["request"]),
        key=lambda c: (len(c["filterAttributes"]), c["kind"] != "TABLE", c["name"]),
    )
    return {
        "candidates": candidates,
        "suggested": viable[0]["id"] if viable else None,
        "basis": "Suggested plans have the fewest post-read filters; ties prefer the base table. This is "
        "a structural comparison, not a measured cost or latency prediction.",
        "proposal": proposed_index(table, request),
        "notes": [
            "Planning uses table metadata only. It does not scan items, execute a query or change indexes.",
            "Returned-attribute selection shrinks the response, not the read units for the same evaluated "
            "items. An index's stored projection can change its entry size and read cost.",
            "Composite string sort keys match from the left: ORDER#2026-09#... supports ORDER# and "
            "ORDER#2026-09 prefixes, not suffix searches. Use consistent UTC timestamps and zero-pad numeric "
            "string components.",
            "One query targets one partition value. Fan-out, multi-attribute native keys and OR conditions "
            "need a different query strategy; this builder never silently turns them into scans.",
        ],
    }
